File: Amazon/TwoSum.js
/**
 * Given an array of integers `nums` and an integer `target`, return indices of the two numbers such that they add up to `target`.
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 * You can return the answer in any order.
 *
 * @param {number[]} nums - The array of integers.
 * @param {number} target - The target sum.
 * @return {number[]} - The indices of the two numbers that add up to the target.
 */
function twoSum(nums, target) {
  const map = new Map();

  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];
    if (map.has(complement)) {
      return [map.get(complement), i];
    }
    map.set(nums[i], i);
  }

  return [];
}

File: Amazon/TwoSum/TwoSum.ex
defmodule TwoSum do
  def two_sum(nums, target) do
    map = Map.new()

    Enum.each_with_index(nums, fn num, index ->
      complement = target - num

      case Map.get(map, complement) do
        nil ->
          Map.put(map, num, index)

        complement_index ->
          return {complement_index, index}
      end
    end)

    nil
  end
end
